packages/vite-plugin-cloudflare/src/websockets.ts

```diff
@@ -0,0 +1,88 @@
+import ws from 'ws';
+import { UNKNOWN_HOST } from './shared';
+import type { Fetcher } from '@cloudflare/workers-types/experimental';
+import type { ReplaceWorkersTypes } from 'miniflare';
+import type { IncomingMessage } from 'node:http';
+import type { Duplex } from 'node:stream';
+import type * as vite from 'vite';
+
+export function handleWebSocket(
+	httpServer: vite.HttpServer,
+	fetcher: ReplaceWorkersTypes<Fetcher>['fetch'],
+	logger: vite.Logger,
+) {
+	const nodeWebSocket = new ws.Server({ noServer: true });
+
+	httpServer.on(
+		'upgrade',
+		async (request: IncomingMessage, socket: Duplex, head: Buffer) => {
+			const url = new URL(request.url ?? '', UNKNOWN_HOST);
+
+			// Ignore Vite HMR WebSockets
+			if (request.headers['sec-websocket-protocol']?.startsWith('vite')) {
+				return;
+			}
+
+			const headers = new Headers();
+
+			for (const [key, value] of Object.entries(request.headers)) {
+				if (typeof value === 'string') {
+					headers.append(key, value);
+				} else if (Array.isArray(value)) {
+					for (const item of value) {
+						headers.append(key, item);
+					}
+				}
+			}
+
+			const response = await fetcher(url, {
+				headers,
+				method: request.method,
+			});
+			const workerWebSocket = response.webSocket;
+
+			if (!workerWebSocket) {
+				socket.destroy();
```

Comment:
why do we need to run `socket.destroy` here? (sorry if it's a silly question, I am not ultra familiar with websockets)

Reply:
well I guess that the worker didn't return a proper socket to use so we can just close the one uses for the upgrade request? 🤔

Reply:
Yep, that's exactly right.